File: 2025-06-16/Boats_To_Save_People.java
package patterns.two_pointers;

import java.util.Arrays;

public class Boats_To_Save_People {
    public static int numRescueBoats(int[] people, int limit) {
        if(people == null || people.length == 0) return 0;

        int n = people.length;
        Arrays.sort(people);

        int boats = 0;
        int left = 0, right = n-1;
        while(left <= right) {
            int sum = people[left] + people[right];
            if(sum <= limit) {
                left++;
                right--;
            }
            else {
                right--;
            }
            boats++;
        }

        return boats;
    }
    public static void main(String[] args) {
        int[] arr = {3,2,2,1};
        System.out.println(numRescueBoats(arr, 3));
    }
}

File: 2026-01-26/Boats_To_Save_People.py
from typing import List


class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        if limit == 0 or not people:
            return 0

        people.sort()
        boat = 0
        l, r = 0, len(people) - 1
        while l <= r:
            s = people[l] + people[r]
            if s <= limit:
                l += 1

            r -= 1
            boat += 1

        return boat


sol = Solution()
print(sol.numRescueBoats([3,5,3,4], 5))